### PR TITLE
Feature - ThongNguyen - CollectionView

### PR DESCRIPTION
**Pull request by mvn-thongnguyen-dn**

### CODE REVIEW

File: 04_View/ThongNT/View/View/CollectionView/BasicCollectionViewController.swift [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        item.backgroundColor = .systemBlue
+        item.label.text = "\(indexPath.item + 1)"
+        return item
+    }
+}
+
+
+final class BasicCollectionViewCell: UICollectionViewCell {
+    let label = UILabel(frame: CGRect(x: 0, y: 0, width: 50, height: 50))
+    override init(frame: CGRect) {
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CollectionView/BasicCollectionViewController.swift:54]: Please use a line spacing to separate

File: 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        _ collectionView: UICollectionView,
+        heightForPhotoAtIndexPath indexPath: IndexPath) -> CGFloat
+}
+
+final class CustomLayout: UICollectionViewLayout {
+    weak var delegate: CustomLayoutDelegate?
+    
+    private let numberOfColumns = 4
+    private let cellPadding: CGFloat = 6
+    private var cache: [UICollectionViewLayoutAttributes] = []
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift:21]: Rename

File: 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    private var contentHeight: CGFloat = 0
+    private var contentWidth: CGFloat {
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift:23]: Separate store properties and computed properties

File: 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        let insets = collectionView.contentInset
+        return collectionView.bounds.width - (insets.left + insets.right)
+    }
+    
+    override var collectionViewContentSize: CGSize {
+        return CGSize(width: contentWidth, height: contentHeight)
+    }
+    
+    override func prepare() {
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift:35]: Should call super.prepare

File: 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        guard cache.isEmpty, let collectionView = collectionView else { return }
+        let columnWidth = contentWidth / CGFloat(numberOfColumns)
+        var xOffset: [CGFloat] = []
+        for column in 0..<numberOfColumns {
+            xOffset.append(CGFloat(column) * columnWidth)
+        }
+        var column = 0
+        var yOffset: [CGFloat] = .init(repeating: 0, count: numberOfColumns)
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift:43]: Rename

File: 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    
+    override var collectionViewContentSize: CGSize {
+        return CGSize(width: contentWidth, height: contentHeight)
+    }
+    
+    override func prepare() {
+        guard cache.isEmpty, let collectionView = collectionView else { return }
+        let columnWidth = contentWidth / CGFloat(numberOfColumns)
+        var xOffset: [CGFloat] = []
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift:38]: Rename

File: 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        return collectionView.bounds.width - (insets.left + insets.right)
+    }
+    
+    override var collectionViewContentSize: CGSize {
+        return CGSize(width: contentWidth, height: contentHeight)
+    }
+    
+    override func prepare() {
+        guard cache.isEmpty, let collectionView = collectionView else { return }
+        let columnWidth = contentWidth / CGFloat(numberOfColumns)
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CollectionView/CustomLayout.swift:37]: Need check numberOfColumns == 0

**Review by mvn-bienle-dn — COMMENTED** [review]
@mvn-thongnguyen-dn Please check my comments.
Thank you!

**Comment by mvn-thongnguyen-dn** [comment]
@mvn-bienle-dn  Thanks for your comments.
Please check my fixes.

**Review by mvn-bienle-dn — APPROVED** [review]
LGTM!

**Pull request merged**